NetworkPolicy: Fix e2e test failures

Kubernetes e2e tests with label `[Feature:NetworkPolicy]` have a couple failures against kube-router master branch. Need to determine if they are legitimate bugs and correct them if so. I will update this issue with how to reproduce the tests.

```
[ 2862.464218] golang[5]: • Failure [51.068 seconds]                                                                                          
[ 2862.464619] golang[5]: [k8s.io] NetworkPolicy                                                                                              
[ 2862.466475] golang[5]: /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:619                
[ 2862.469078] golang[5]:   should support allow-all policy [Feature:NetworkPolicy] [It]                                                      
[ 2862.471207] golang[5]:   /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network_policy.go:270                   
[ 2862.471869] golang[5]:   checking client-a could communicate with server.                                                                  
[ 2862.473288] golang[5]:   Expected error:                                                                                                   
[ 2862.473798] golang[5]:       <*errors.errorString | 0xc4210dc8e0>: {                                                                       
[ 2862.476920] golang[5]:           s: "pod \"client-a\" failed with status: {Phase:Failed Conditions:[{Type:Initialized Status:True LastProbe
Time:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-08-08 17:31:02 +0000 UTC Reason: Message:} {Type:Ready Status:False LastProbeTime:0
001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-08-08 17:31:05 +0000 UTC Reason:ContainersNotReady Message:containers with unready status
: [client-a-container]} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-08-08 17:31:02 +000
0 UTC Reason: Message:}] Message: Reason: HostIP:172.17.4.201 PodIP: StartTime:2017-08-08 17:31:02 +0000 UTC InitContainerStatuses:[] Containe
rStatuses:[{Name:client-a-container State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:1,Signal:0,Reason:Error,Messa
ge:,StartedAt:2017-08-08 17:31:03 +0000 UTC,FinishedAt:2017-08-08 17:31:04 +0000 UTC,ContainerID:docker://a5c934887353ef59a04058ad5800ac0a1a72
e24b3ba65210b2b66001fd7d351b,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:gcr.io/google_c
ontainers/redis:e2e ImageID:docker-pullable://gcr.io/google_containers/redis@sha256:f066bcf26497fbc55b9bf0769cb13a35c0afa2aa42e737cc46b7fb04b2
3a2f25 ContainerID:docker://a5c934887353ef59a04058ad5800ac0a1a72e24b3ba65210b2b66001fd7d351b}] QOSClass:BestEffort}",                         
[ 2862.481588] golang[5]:       }                                                                                                             
[ 2862.482472] golang[5]:       pod "client-a" failed with status: {Phase:Failed Conditions:[{Type:Initialized Status:True LastProbeTime:0001-
01-01 00:00:00 +0000 UTC LastTransitionTime:2017-08-08 17:31:02 +0000 UTC Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-08-08 17:31:05 +0000 UTC Reason:ContainersNotReady Message:containers with unready status: [client$a-container]} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-08-08 17:31:02 +0000 UTC Rea$on: Message:}] Message: Reason: HostIP:172.17.4.201 PodIP: StartTime:2017-08-08 17:31:02 +0000 UTC InitContainerStatuses:[] ContainerStatuses$[{Name:client-a-container State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:1,Signal:0,Reason:Error,Message:,Start$dAt:2017-08-08 17:31:03 +0000 UTC,FinishedAt:2017-08-08 17:31:04 +0000 UTC,ContainerID:docker://a5c934887353ef59a04058ad5800ac0a1a72e24b3ba65$10b2b66001fd7d351b,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:gcr.io/google_containers$redis:e2e ImageID:docker-pullable://gcr.io/google_containers/redis@sha256:f066bcf26497fbc55b9bf0769cb13a35c0afa2aa42e737cc46b7fb04b23a2f25 Co$tainerID:docker://a5c934887353ef59a04058ad5800ac0a1a72e24b3ba65210b2b66001fd7d351b}] QOSClass:BestEffort}                                     
[ 2862.483490] golang[5]:   not to have occurred                       
[ 2862.486790] golang[5]:   /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network_policy.go:341                   
[ 2862.488252] golang[5]: ------------------------------               
[ 2862.491128] golang[5]: SSSSSSSSSSSSSSSSSSSSSSSSSSSSAug  8 17:31:47.544: INFO: Running AfterSuite actions on all node                       
[ 2862.496176] golang[5]: Aug  8 17:31:47.544: INFO: Running AfterSuite actions on node 1                                                     
[ 2862.498761] golang[5]: Summarizing 2 Failures:                      
[ 2862.502385] golang[5]: [Fail] [k8s.io] NetworkPolicy [It] should enforce policy based on NamespaceSelector [Feature:NetworkPolicy]         
[ 2862.504577] golang[5]: /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network_policy.go:356                     
[ 2862.507136] golang[5]: [Fail] [k8s.io] NetworkPolicy [It] should support allow-all policy [Feature:NetworkPolicy]                          
[ 2862.510478] golang[5]: /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network_policy.go:341                     [ 2862.524102] golang[5]: Ran 6 of 652 Specs in 492.907 seconds        
[ 2862.525975] golang[5]: FAIL! -- 4 Passed | 2 Failed | 0 Pending | 646 Skipped --- FAIL: TestE2E (492.96s)                                  [ 2862.526576] golang[5]: FAIL     
[ 2862.527053] golang[5]: Ginkgo ran 1 suite in 8m13.509020038s        
[ 2862.527547] golang[5]: Test Suite Failed
```

--- a/app/controllers/network_policy_controller.go
+++ b/app/controllers/network_policy_controller.go
@@ -20,6 +20,7 @@ import (
 	"github.com/golang/glog"
 	"github.com/janeczku/go-ipset/ipset"
 	"k8s.io/client-go/kubernetes"
+	api "k8s.io/client-go/pkg/api/v1"
 	apiv1 "k8s.io/client-go/pkg/api/v1"
 	apiextensions "k8s.io/client-go/pkg/apis/extensions/v1beta1"
 	networking "k8s.io/client-go/pkg/apis/networking/v1"
@@ -252,6 +253,12 @@ func (npc *NetworkPolicyController) syncNetworkPolicyChains() (map[string]bool, 
 			return nil, fmt.Errorf("Failed to run iptables command: %s", err.Error())
 		}
 
+		// From network policy spec: "If field 'Ingress' is empty then this NetworkPolicy does not allow any traffic "
+		// so no whitelist rules to be added to the network policy
+		if policy.ingressRules == nil {
+			continue
+		}
+
 		// run through all the ingress rules in the spec and create iptable rules
 		// in the chain for the network policy
 		for i, ingressRule := range policy.ingressRules {
@@ -325,13 +332,6 @@ func (npc *NetworkPolicyController) syncNetworkPolicyChains() (map[string]bool, 
 			// case where nether ports nor from details are speified in the ingress rule
 			// so match on all ports, protocol, source IP's
 			if ingressRule.matchAllSource && ingressRule.matchAllPorts {
-
-				// if no ports or source information is present in spec this is specical case
-				// where network policy does not allow any traffic
-				if npc.v1NetworkPolicy {
-					continue
-				}
-
 				comment := "rule to ACCEPT traffic from source pods to dest pods selected by policy name: " +
 					policy.name + " namespace " + policy.namespace
 				args := []string{"-m", "comment", "--comment", comment,
@@ -660,7 +660,6 @@ func buildNetworkPoliciesInfo() (*[]networkPolicyInfo, error) {
 		}
 		matchingPods, err := watchers.PodWatcher.ListByNamespaceAndLabels(policy.Namespace, policy.Spec.PodSelector.MatchLabels)
 		newPolicy.destPods = make(map[string]podInfo)
-		newPolicy.ingressRules = make([]ingressRule, 0)
 		if err == nil {
 			for _, matchingPod := range matchingPods {
 				newPolicy.destPods[matchingPod.Status.PodIP] = podInfo{ip: matchingPod.Status.PodIP,
@@ -668,6 +667,12 @@ func buildNetworkPoliciesInfo() (*[]networkPolicyInfo, error) {
 					namespace: matchingPod.ObjectMeta.Namespace,
 					labels:    matchingPod.ObjectMeta.Labels}
 			}
+		}
+
+		if policy.Spec.Ingress == nil {
+			newPolicy.ingressRules = nil
+		} else {
+			newPolicy.ingressRules = make([]ingressRule, 0)
 		}
 
 		for _, specIngressRule := range policy.Spec.Ingress {
@@ -693,8 +698,26 @@ func buildNetworkPoliciesInfo() (*[]networkPolicyInfo, error) {
 				ingressRule.matchAllSource = true
 			} else {
 				ingressRule.matchAllSource = false
+				var matchingPods []*api.Pod
+				var err error
 				for _, peer := range specIngressRule.From {
-					matchingPods, err := watchers.PodWatcher.ListByNamespaceAndLabels(policy.Namespace, peer.PodSelector.MatchLabels)
+					// spec must have either of PodSelector or NamespaceSelector
+					if peer.PodSelector != nil {
+						matchingPods, err = watchers.PodWatcher.ListByNamespaceAndLabels(policy.Namespace,
+							peer.PodSelector.MatchLabels)
+					} else if peer.NamespaceSelector != nil {
+						namespaces, err := watchers.NamespaceWatcher.ListByLabels(peer.NamespaceSelector.MatchLabels)
+						if err != nil {
+							return nil, errors.New("Failed to build network policies info due to " + err.Error())
+						}
+						for _, namespace := range namespaces {
+							namespacePods, err := watchers.PodWatcher.ListByNamespaceAndLabels(namespace.Name, nil)
+							if err != nil {
+								return nil, errors.New("Failed to build network policies info due to " + err.Error())
+							}
+							matchingPods = append(matchingPods, namespacePods...)
+						}
+					}
 					if err == nil {
 						for _, matchingPod := range matchingPods {
 							ingressRule.srcPods = append(ingressRule.srcPods,

--- a/app/watchers/namespace_watcher.go
+++ b/app/watchers/namespace_watcher.go
@@ -10,6 +10,8 @@ import (
 	"k8s.io/client-go/kubernetes"
 	api "k8s.io/client-go/pkg/api/v1"
 	cache "k8s.io/client-go/tools/cache"
+	listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
 )
 
 type NamespaceUpdate struct {
@@ -66,6 +68,16 @@ func (nsw *namespaceWatcher) List() []*api.Namespace {
 		namespace_instances[i] = ins.(*api.Namespace)
 	}
 	return namespace_instances
+}
+
+func (nsw *namespaceWatcher) ListByLabels(set labels.Set) ([]*api.Namespace, error) {
+	namespaceLister := listers.NewNamespaceLister(nsw.namespaceLister)
+	matchedNamespaces, err := namespaceLister.List(set.AsSelector())
+	if err != nil {
+		return nil, err
+	} else {
+		return matchedNamespaces, nil
+	}
 }
 
 func (nsw *namespaceWatcher) RegisterHandler(handler NamespaceUpdatesHandler) {

